File: new/atoms/nedarim-auto-match-charges.mjs
/** חוט · nedarim-auto-match-charges — Golden. חוזה: nedarim-auto-match-charges.contract.md
 * מוצא: nedarimSync.ts:324 (autoMatchCharges) + keysOf:87 (inline). חוק-4 verbatim.
 * שיוך-אוטומטי מרובה: אינדקס-מפתחות O(S) ואז מפתח-חזק-ביותר לכל עסקה O(M). שם-בלבד לא נכלל.
 * שקעים: normId,normPhone,normSearch.
 */
export function autoMatchCharges(charges, supporters, { normId, normPhone, normSearch }) {
  const keysOf = (o) => {
    const ks = [];
    const ext = (o.extId || '').trim(); if (ext) ks.push('ext:' + ext);
    const id = normId(o.idNum || o.zeout); if (id) ks.push('id:' + id);
    for (const p of [o.phone, o.phone2, o.phone3]) { const ph = normPhone(p || ''); if (ph.length >= 7) ks.push('ph:' + ph); }
    const em = (o.email || '').trim().toLowerCase(); if (em) ks.push('em:' + em);
    const n = normSearch(o.name || ''); const c = normSearch(o.city || ''); if (n && c) ks.push('nc:' + n + '|' + c);
    return ks;
  };
  const idx = new Map();
  for (const sp of supporters) {
    for (const k of keysOf({ extId: sp.extId, idNum: sp.idNum, phone: sp.phone, email: sp.email })) {
      if (!idx.has(k)) idx.set(k, sp.id);
    }
  }
  const out = [];
  for (const c of charges) {
    let supId;
    for (const k of keysOf({ extId: c.toremId, zeout: c.zeout, phone: c.phone, email: c.email })) {
      const hit = idx.get(k);
      if (hit) { supId = hit; break; }
    }
    if (supId) out.push({ supId, charge: c });
  }
  return out;
}

File: new/dart-maor/nedarim-auto-match-charges.dart
// ⚛️ אטום-Dart (דרגת-חוזה) · autoMatchCharges — שיוך-אצווה יעיל: אינדקס-מפתחות O(S)
// + מפתח-חזק פר-עסקה; שם-בלבד לא נכלל.
// מוצא: maor-system/src/lib/nedarimSync.ts:324 + keysOf:87 (inline) · המקור: new/atoms/nedarim-auto-match-charges.mjs.
// טוהר: פונקציית top-level עצמאית, אפס import (dart:core בלבד). חוק-4 — זהה-ביט למקור-JS.
//        שקעים (חוק-1/חוק-3): normId, normPhone, normSearch — פרמטרים-נקובים.
//
// הערות-המרה (JS→Dart):
//  • `new Map()` + `idx.has/set/get` ⇒ Map (LinkedHashMap). מפתח ראשון-מנצח (`if(!idx.has(k))`).
//  • `if(hit){supId=hit;break;}` — הראשון-שנמצא ⇒ עצירה. `if(supId)` ⇒ hit truthy.
//  • הפלט List של Map בסדר supId → charge.

bool _truthy(dynamic v) {
  if (v == null || v == false) return false;
  if (v is String) return v.isNotEmpty;
  if (v is num) return v != 0 && !v.isNaN;
  return true;
}

dynamic _or(dynamic a, dynamic b) => _truthy(a) ? a : b;

String _s(dynamic v) => v is String ? v : '';

/// Efficient batch matching. Verbatim port of nedarim-auto-match-charges.mjs.
List<Map<String, dynamic>> autoMatchCharges(
  List charges,
  List supporters, {
  required String Function(String?) normId,
  required String Function(String?) normPhone,
  required String Function(dynamic) normSearch,
}) {
  List<String> keysOf(Map o) {
    final ks = <String>[];
    final ext = _s(o['extId']).trim();
    if (ext.isNotEmpty) ks.add('ext:$ext');
    final id = normId(_or(o['idNum'], o['zeout']) as String?);
    if (id.isNotEmpty) ks.add('id:$id');
    for (final p in [o['phone'], o['phone2'], o['phone3']]) {
      final ph = normPhone(_or(p, '') as String?);
      if (ph.length >= 7) ks.add('ph:$ph');
    }
    final em = (_or(o['email'], '') as String).trim().toLowerCase();
    if (em.isNotEmpty) ks.add('em:$em');
    final n = normSearch(_or(o['name'], ''));
    final c = normSearch(_or(o['city'], ''));
    if (n.isNotEmpty && c.isNotEmpty) ks.add('nc:$n|$c');
    return ks;
  }

  final idx = <String, dynamic>{};
  for (final sp in supporters) {
    for (final k in keysOf({
      'extId': sp['extId'],
      'idNum': sp['idNum'],
      'phone': sp['phone'],
      'email': sp['email'],
    })) {
      if (!idx.containsKey(k)) idx[k] = sp['id'];
    }
  }
  final out = <Map<String, dynamic>>[];
  for (final c in charges) {
    dynamic supId;
    for (final k in keysOf({
      'extId': c['toremId'],
      'zeout': c['zeout'],
      'phone': c['phone'],
      'email': c['email'],
    })) {
      final hit = idx[k];
      if (_truthy(hit)) {
        supId = hit;
        break;
      }
    }
    if (_truthy(supId)) out.add({'supId': supId, 'charge': c});
  }
  return out;
}
